iothub/device/src/Transport/AmqpIoT/AmqpIoTReceivingLink.cs

```diff
@@ -257,89 +257,98 @@ private void DisposeDelivery(AmqpMessage amqpMessage, bool settled, Accepted acc
 
         internal void RegisterTwinListener(Action<Twin, string, TwinCollection> onDesiredPropertyReceived)
         {
-            _onDesiredPropertyReceived = onDesiredPropertyReceived;
-            _receivingAmqpLink.RegisterMessageListener(OnDesiredPropertyReceived);
+            _onTwinMessageReceived = onDesiredPropertyReceived;
+            _receivingAmqpLink.RegisterMessageListener(OnTwinChangesReceived);
         }
 
-        private void OnDesiredPropertyReceived(AmqpMessage amqpMessage)
+        private void OnTwinChangesReceived(AmqpMessage amqpMessage)
         {
             if (Logging.IsEnabled)
             {
-                Logging.Enter(this, amqpMessage, $"{nameof(OnDesiredPropertyReceived)}");
+                Logging.Enter(this, amqpMessage, $"{nameof(OnTwinChangesReceived)}");
             }
 
             try
             {
                 _receivingAmqpLink.DisposeDelivery(amqpMessage, true, AmqpIoTConstants.AcceptedOutcome);
                 string correlationId = amqpMessage.Properties?.CorrelationId?.ToString();
-
-                if (!VerifyResponseMessage(amqpMessage))
-                {
-                    _onDesiredPropertyReceived.Invoke(null, correlationId, null);
-                }
+                int status = GetStatus(amqpMessage);
 
                 Twin twin = null;
                 TwinCollection twinProperties = null;
 
-                if (correlationId != null)
+                if (status >= 400)
                 {
-                    if (amqpMessage.BodyStream != null)
-                    {
-                        // This a result of a GET TWIN so return (set) the full twin
-                        using (StreamReader reader = new StreamReader(amqpMessage.BodyStream, System.Text.Encoding.UTF8))
-                        {
-                            string body = reader.ReadToEnd();
-                            var properties = JsonConvert.DeserializeObject<TwinProperties>(body);
-                            twin = new Twin(properties);
-                        }
-                    }
-                    else
+                    _onTwinMessageReceived.Invoke(null, correlationId, null);
+                    if (correlationId.StartsWith(AmqpTwinMessageType.Get.ToString(), StringComparison.OrdinalIgnoreCase))
                     {
-                        // This is a desired property ack from the service
-                        twin = new Twin();
+                        // Handle failures
+                        _onTwinMessageReceived.Invoke(null, correlationId, null);
+                        string error = null;
+                        var reader = new StreamReader(amqpMessage.BodyStream, System.Text.Encoding.UTF8);
```

Comment:
using instead of explicit call to dispose (if an exception were to occur, it would not be disposed)